pkg/threescale/connectors/backend/local_cache.go

```diff
@@ -58,15 +59,68 @@ func (l LocalCache) Set(cacheKey string, cv CacheValue) {
 	l.ds.Set(cacheKey, cv)
 }
 
+// Report cached entries to 3scale backend
 func (l LocalCache) Report() {
 	ticker := time.NewTicker(l.reporter.interval)
+	// note - it would be nice if these were exposed as iota's to avoid doing this
+	ascendingPeriodSequence := []client.LimitPeriod{client.Minute, client.Hour, client.Day, client.Month, client.Eternity}
 
 	go func() {
 		for {
 			select {
 			case <-ticker.C:
 				l.reporter.cache.IterCb(func(key string, v interface{}) {
-					//TODO - walk the cache and handle the reporting
+					reportMetrics := client.Metrics{}
+					cachedValue := v.(CacheValue)
+
+					for reportableMetric, _ := range cachedValue.LastResponse.GetUsageReports() {
+						for localMetric, limitMap := range cachedValue.LimitCounter {
+							if reportableMetric != localMetric {
+								// we dont need to deal with this particular metric since it was learned and incremented as part of the hierarchy tree
+								continue
+							}
+							for _, ascendingPeriod := range ascendingPeriodSequence {
+								if lowestPeriod, ok := limitMap[ascendingPeriod]; ok {
+									reportMetrics.Add(reportableMetric, lowestPeriod.current)
+									break
+								}
+							}
+						}
+					}
```

Comment:
I think we should be iterating the whole data set we have rather than the usage reports from the last authorization, because we could have metrics which are not directly limited (thus not in the usage reports) but indirectly limited via a hierarchy (ie. we could think of a case in which a method `orders` is not being limited directly, but indirectly because it is a children of `Hits` which would be limited and appearing in the usage reports).

Note that here we could be over-reporting. We need to figure out which metrics we just contributed to by directly reporting (adding to the current value because we matched on the metric, even though they might have no limit) vs indirectly via hierarchy (ie. we didnt directly match on the metric, but we added to it because a children was matched directly), and report only direct hits because 3scale will add the indirect hits itself (we might consider adding some extension to avoid this double computation in the cache and 3scale in the future). For example:

`Hits` being limited and then having calls that match it directly, but also calls that match just a child of it directly (and indirectly add to `Hits`). We need to make this distinction. One option would be to keep two counters, but that would complicate the checks in the hot path, so a better option would be to walk the full set of metrics we have had with activity this way:

> For each metric we have data, subtract the previous current value, and subtract that value to its parent (and grandparent, grand-grandparent, etc if existing). The report should then include directly any metric that is non-zero.

This way, if we have two metrics, parent and child, with current values 10 and 5 respectively, and we received two requests, one matching child with 1 hit added, and another matching parent with 1 hit added:

1. We would add 1 to child and 1 to parent after the first request => child: 6, parent: 11 (and we'd check limits if present on both).
2. We would add 1 to parent after the second request => child: 6, parent: 12 (check limits, etc).
3. At report time, we'd find `child` and subtract its previous current value, 5 so that => child: 1, then follow the hierarchy and subtract 5 to its parent, parent: 7.
4. Still at report time, we'd find `parent` and subtract its previous current value, 10 so that => parent: 2.
5. Once we are done with this loop, we got the actual delta for this reporting iteration, but we can't push that to 3scale because it will use the hierarchy, so we go over the metrics again and look for metrics that have 1 or more children.
6. Since `child` has no children, we don't do anything to it.
7. Since `parent` has `child` as children, we *look first* at `child`, and if it had children we'd first go look at them before proceeding. For every child, we take their remaining value and subtract it from the parent. Since `parent` only has one child, `child`, and that child has been processed already (it has 0 children), we subtract `child`'s value from `parent`, so that => parent: 1.
8. Once this is done for all metrics, we get the final set of metrics and value to report: `child: 1, parent: 1`. (metrics which end up with a value of 0 don't need to be reported)